src/include/brain/brain.h

```diff
@@ -15,52 +15,88 @@
-  // TODO(tianyu): Extend this interface for richer interaction
-  virtual void RunJob(BrainEnvironment *) = 0;
+  explicit BrainJob(BrainEnvironment *env) : env_(env) {}
+
+  virtual ~BrainJob() = default;
+
+  // This is separate from the user-defined OnJobInvocation to allow for better
+  // interfacing with the libevent API.
+  /**
+   * Invokes this job to be run. The brain framework will call this method.
+   *
+   */
+  inline void Invoke() { OnJobInvocation(env_); }
+  // TODO(tianyu): Extend this interface for richer behavior
+  /**
+   * Executed as the main body of the job, filled in by the user. Use the
+   * provided BrainEnvironment for interaction with Brain's resources.
+   */
+  virtual void OnJobInvocation(BrainEnvironment *) = 0;
+ private:
+  BrainEnvironment *env_;
 };
 
-class ExampleBrainJob: public BrainJob {
+/**
+ * Simple implementation of a BrainJob.
+ */
+class SimpleBrainJob : public BrainJob {
  public:
-   void RunJob(BrainEnvironment *) override {
-     // TODO(tianyu): Replace with real address
-     capnp::EzRpcClient client("localhost:15445");
-     PelotonService::Client peloton_service = client.getMain<PelotonService>();
-     auto request = peloton_service.createIndexRequest();
-     request.getRequest().setIndexKeys(42);
-     auto response = request.send().wait(client.getWaitScope());
-   }
+  explicit SimpleBrainJob(BrainEnvironment *env,
+                          std::function<void(BrainEnvironment *)> task)
+      : BrainJob(env), task_(std::move(task)) {}
+  inline void OnJobInvocation(BrainEnvironment *env) override { task_(env); }
+ private:
+  std::function<void(BrainEnvironment *)> task_;
 };
 
+/**
+ * Main running component of the brain. Events can be registered on this event
+ * loop and once Run is called, it will invoke handlers every specified time
+ * interval
+ */
 class Brain {
  public:
   // TODO(tianyu): Add necessary parameters to initialize the brain's resources
   Brain() : scheduler_(0) {}
 
+  ~Brain() {
+    for (auto entry : jobs_)
+      delete entry.second;
+  }
+
+  template <typename BrainJob, typename... Args>
   inline void RegisterJob(const struct timeval *period,
-                          std::string name,
-                          BrainJob &job) {
-    auto callback = [](int, short, void *pair) {
-      auto *job_env_pair = reinterpret_cast<std::pair<BrainJob *, BrainEnvironment *> *>(pair);
-      job_env_pair->first->RunJob(job_env_pair->second);
+                          std::string name, Args... args) {
+    auto *job = new BrainJob(&env_, args...);
+    jobs_[name] = job;
+    auto callback = [](int, short, void *arg) {
+      reinterpret_cast<BrainJob *>(arg)->Invoke();
     };
-    // TODO(tianyu) Deal with this memory
-    auto *pair = new std::pair<BrainJob *, BrainEnvironment *>(&job, &env_);
     job_handles_[name] =
-        scheduler_.RegisterPeriodicEvent(period, callback, pair);
+        scheduler_.RegisterPeriodicEvent(period, callback, job);
```

Comment:
As far as I understood, this "callback" here is just a wrapper to invoke the brain job, right? Then why is called a "callback"? The naming is just a bit confusing to me because usually, the callback contains the extra work you need to do after the invocation returns.

Reply:
This is libevent parlance. (I agree that really we should call them handlers, but since they call it callback...)